### DISCUSSION TITLE
Expose underline safe crate as part of an official target

### DISCUSSION THREAD

**Discussion by Urgau** [sunfishcode/mustang#50] [Ideas]
One thing that would be interesting, would be to expose a safe crate (`rsix` maybe ?) to all program that use our official target.

Something like `extern crate rsix;` at the top of the program would give programs access to (mostly) safe abstraction, avoiding the use for `libc`/`nix`, although they should work.

What is everyone take on that ?

**Comment by sunfishcode** [comment]
My first take here is that much of `std` has been designed with somewhat higher-level APIs which are more portable to Windows and other platforms, and that most general-purpose Rust code should ideally continue to use those APIs where applicable. As such, it seems ok that lower-level APIs aren't exposed by default.

However, one thing I don't know is that if std were to depend on rsix, and users separately depend on rsix, will we end up with two copies of rsix linked into programs? If so, it might be worth looking into what we could do to avoid that.

Another thing to consider is, what are the use cases? I'm tending to think in terms of starting with general-purpose Rust code that isn't specifically written for Mustang, however I'm interested in how others see Mustang being used in practice.

**Reply by bjorn3** [reply]
> However, one thing I don't know is that if std were to depend on rsix, and users separately depend on rsix, will we end up with two copies of rsix linked into programs?

Yes, they will. The sysroot is compiled such that crates libstd depends on don't conflict with identical crates compiled by the user. They will just be both included.

**Reply by sunfishcode** [reply]
Ok. Well, `rsix` is pretty thin; most functions can be fully inlined, so maybe we can wait and see what the impact is when we get there.
